homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,160 @@
+"""
+Support for Somfy hubs.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/somfy/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant import config_entries
+from homeassistant.components.somfy import config_flow
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import Throttle
+
+API = 'api'
+
+DEVICES = 'devices'
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'somfy'
+
+CONF_CLIENT_ID = 'client_id'
+CONF_CLIENT_SECRET = 'client_secret'
+
+SOMFY_AUTH_CALLBACK_PATH = '/auth/somfy/callback'
+SOMFY_AUTH_START = '/auth/somfy'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_CLIENT_SECRET): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SOMFY_COMPONENTS = ['cover']
+
+
+async def async_setup(hass, config):
+    """Set up the Somfy component."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = {}
+
+    config_flow.register_flow_implementation(
+        hass, config[DOMAIN][CONF_CLIENT_ID],
+        config[DOMAIN][CONF_CLIENT_SECRET])
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+        ))
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Somfy from a config entry."""
+    def token_saver(token):
+        _LOGGER.debug('Saving updated token')
+        entry.data[CONF_TOKEN] = token
+        hass.config_entries.async_update_entry(entry, data={**entry.data})
+
+    # Force token update.
+    from pymfy.api.somfy_api import SomfyApi
+    hass.data[DOMAIN][API] = SomfyApi(
```

Comment:
`hass.data[DOMAIN]` is only defined if `DOMAIN in config`

Reply:
Do you talk about a use case where the end user setup a first time the component, then for any reason remove it?

Reply:
It might not be possible with the current use case, but I am talking about the use case where `DOMAIN` is not in the config. I think that you should always set `hass.data[DOMAIN] = {}` inside async_setup.

Reply:
This works fine (don't forget `})`):

```suggestion
    hass.data.setdefault(DOMAIN, {}).update({API: SomfyApi(
```

Reply:
Look at https://github.com/home-assistant/home-assistant/pull/19548/files#diff-5f6fcaa13fcf30fc2741aedd6cc2666bR53. I think it's what you are talking about.

Reply:
Mine works fine too :)

Reply:
@balloob I tried your use case. There was an issue indeed but not here. See my commit: https://github.com/home-assistant/home-assistant/pull/19548/commits/d388b48fb69a32440e49413c6365ab3863d8c1fd#diff-05a80de0a3ff7f74840c782130a44750R51
So now, if the config is missing, code is protected at startup or when end user trigger manually the integration.